### PR TITLE
[quick_actions] Android handle quick action without restart

### PR DESCRIPTION
**Pull request by TabooSun**
In Android, support app opening (triggered by quick_actions) without restarting the entire app.

Fixes https://github.com/flutter/flutter/issues/83251

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
@blasten to review or delegate

**Comment by stuartmorgan** [comment]
Adding @camsim99 for Android review, per the new owners mapping.

File: packages/quick_actions/quick_actions/CHANGELOG.md
Comment (camsim99) [inline packages/quick_actions/quick_actions/CHANGELOG.md:5]: @stuartmorgan Should this include the `NEXT` item since this is a version bump?

File: packages/quick_actions/quick_actions_android/android/src/main/AndroidManifest.xml
Comment (camsim99) [inline packages/quick_actions/quick_actions_android/android/src/main/AndroidManifest.xml:1]: Why remove this line?

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (camsim99) [inline packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java:60]: I appreciate the organization of the tests, but can you remove the "arrange", "act", and "assert" comments? I think the clarity will remain!

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (camsim99) [inline packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java:67]: I may be missing some context here, but can you just use the `ShortcutInfo`s to test versus creating this new `Shortcut` object?

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/Shortcut.java
Comment (camsim99) [inline packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/Shortcut.java:10]: Made a comment about this above, but do we need this object for testing?

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (TabooSun) [reply packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java:67]: I meant to keep the test variants created from `createMockShortcuts` as simple as possible, without the need to go through the complexity of creating a `ShortcutInfo`.

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (TabooSun) [reply packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java:60]: Sure.

File: packages/quick_actions/quick_actions_android/android/src/main/AndroidManifest.xml
Comment (TabooSun) [reply packages/quick_actions/quick_actions_android/android/src/main/AndroidManifest.xml:1]: Ops, it was removed by my IDE while I commit it.

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (camsim99) [reply packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java:65]: What is that complexity? If doable, I would suggest testing with `ShortcutInfo`s for clarity since this is the object actually used in the plugin.

Comment (TabooSun) [reply packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java]: I have updated the code.

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (camsim99) [inline packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java:48]: I didn't notice this before, but good catch!

**Review by camsim99 — APPROVED** [review]
LGTM! Just need to fix the failing tests and second review. @stuartmorgan can you take a look or verify the version bump?

File: packages/quick_actions/quick_actions/CHANGELOG.md
Comment (stuartmorgan) [reply packages/quick_actions/quick_actions/CHANGELOG.md:6]: Yes, but the larger problem is that this version bump is to the wrong package. The changes are to `quick_actions_android`, not `quick_actions`, so that's what needs CHANGELOG and version updates.

@TabooSun Please make sure the `publishable` CI check is passing when you update your PR; the failure messages there explain the problem.

File: packages/quick_actions/quick_actions/pubspec.yaml
Comment (stuartmorgan) [inline packages/quick_actions/quick_actions/pubspec.yaml:6]: The way Dart versioning works for pre-1.0, this is a breaking version change (see link in the PR checklist about Dart versions). Is there anything about this change that actually breaks clients?

File: packages/quick_actions/quick_actions_android/android/src/main/java/io/flutter/plugins/quickactions/QuickActionsPlugin.java
Comment (stuartmorgan) [inline packages/quick_actions/quick_actions_android/android/src/main/java/io/flutter/plugins/quickactions/QuickActionsPlugin.java:77]: Why are calling Java code from Java code via the method channel? And why are you calling a getter but not using the result?

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (stuartmorgan) [inline packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java:6]: This violates style: https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (stuartmorgan) [inline packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java:17]: Same.

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (stuartmorgan) [inline packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java:2]: Please revert this change; it's why the format check is failing. The licenses follow an exact format, whitespace included.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

**Comment by stuartmorgan** [comment]
@camsim99 Before this lands, you'll want to push some trivial change to the PR so that there's a test run triggered by you; by policy we can't run FTL tests for pushes from non-project-members, and that's currently the only thing that runs Android integration tests (like the ones added here), so we should validate that they pass in CI by triggering a run ourselves.

File: packages/quick_actions/quick_actions_android/android/src/main/java/io/flutter/plugins/quickactions/QuickActionsPlugin.java
Comment (TabooSun) [reply packages/quick_actions/quick_actions_android/android/src/main/java/io/flutter/plugins/quickactions/QuickActionsPlugin.java]: I want to reuse the logic in `getLaunchAction`. 

Since this seems to be a bad idea, I have tried another way that is exposing a public method from `MethodCallHandlerImpl`. Do let me know if this is okay.

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (TabooSun) [reply packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java]: Done.

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (TabooSun) [reply packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java]: Done.

**Comment by stuartmorgan** [comment]
I see that there have been changes pushed here; is this ready for another review?

**Comment by TabooSun** [comment]
> I see that there have been changes pushed here; is this ready for another review?

Yes.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/quick_actions/quick_actions/CHANGELOG.md
Comment (stuartmorgan) [reply packages/quick_actions/quick_actions/CHANGELOG.md]: I went ahead and fix this while resolving conflicts with `main`.

**Comment by stuartmorgan** [comment]
I've pushed a commit, so we'll get a CI run of the new tests and make sure they are working there.

**Comment by stuartmorgan** [comment]
@TabooSun This PR isn't passing tests in CI:

appShortcutExistsAfterLongPressingAppIcon
```
java.lang.AssertionError: The specified shortcut label 'Action two' does not exist.
at org.junit.Assert.fail(Assert.java:88)
at org.junit.Assert.assertTrue(Assert.java:41)
at io.flutter.plugins.quickactionsexample.QuickActionsTest.appShortcutExistsAfterLongPressingAppIcon(QuickActionsTest.java:94)
...
```

appShortcutsAreCreated
```
java.lang.AssertionError: expected:<2> but was:<1>
at org.junit.Assert.fail(Assert.java:88)
at org.junit.Assert.failNotEquals(Assert.java:834)
at org.junit.Assert.assertEquals(Assert.java:645)
at org.junit.Assert.assertEquals(Assert.java:631)
at io.flutter.plugins.quickactionsexample.QuickActionsTest.appShortcutsAreCreated(QuickActionsTest.java:75)
```

appShortcutLaunchActivityAfterPressing
```
java.lang.AssertionError: AppShortcut:action_one does not launch the correct activity
at org.junit.Assert.fail(Assert.java:88)
at org.junit.Assert.assertTrue(Assert.java:41)
at io.flutter.plugins.quickactionsexample.QuickActionsTest.appShortcutLaunchActivityAfterPressing(QuickActionsTest.java:118)
```

**Comment by TabooSun** [comment]
Since all the tests have passed, @stuartmorgan feel free to review.

**Comment by stuartmorgan** [comment]
> Since all the tests have passed, @stuartmorgan feel free to review.

They don't pass, they just didn't run (see [my comment above](https://github.com/flutter/plugins/pull/5048#issuecomment-1116444829)). They are still failing when I push.

Do these tests pass for you locally?

**Comment by TabooSun** [comment]
Unfortunately, the error you provided isn't helpful enough; the only info I can extract from the error is one of the shortcuts isn't being created. Looking at https://github.com/flutter/plugins/blob/39020ce73c7e7070e30f3ebcf345492441b477eb/packages/quick_actions/quick_actions_android/android/src/main/java/io/flutter/plugins/quickactions/MethodCallHandlerImpl.java#L83, I am wondering if there is such error in the log.

I can try help diagnose further if you provide me more detail or full log.

Yes, they pass for me locally, if the locally you mean here is running the test against all of my own devices instead of the devices on Firebase Test Lab.

**Comment by TabooSun** [comment]
@stuartmorgan Thanks for the info. I have updated the code to ensure "actions ready" appears.

**Comment by stuartmorgan** [comment]
The test failures are still the same, so apparently that's not the issue.

I recommend building [the same way the FTL tests do](https://console.firebase.google.com/project/flutter-cirrus/testlab/histories/bh.38154836b3d1fd98/matrices/5519422818441126319) locally, and seeing if you can reproduce it that way.

**Comment by TabooSun** [comment]
@stuartmorgan I have tested in local and fixed the integration test. Could you help rerun?

**Comment by stuartmorgan** [comment]
Looks like maybe there's some flake now? The failures are only on master, and differ between the two devices. The API 26 device had two instance of:
```
androidx.test.uiautomator.UiObjectNotFoundException: UiSelector[SCROLLABLE=true]
at androidx.test.uiautomator.UiScrollable.scrollForward(UiScrollable.java:417)
at androidx.test.uiautomator.UiScrollable.scrollForward(UiScrollable.java:398)
at io.flutter.plugins.quickactionsexample.QuickActionsTest.findAppIcon(QuickActionsTest.java:169)
```

one in `appShortcutLaunchActivityAfterPressing` and one in `appShortcutExistsAfterLongPressingAppIcon`, while the API 30 device had:
```
java.lang.AssertionError: AppShortcut:action_one does not launch the correct activity
at org.junit.Assert.fail(Assert.java:88)
at org.junit.Assert.assertTrue(Assert.java:41)
at io.flutter.plugins.quickactionsexample.QuickActionsTest.appShortcutLaunchActivityAfterPressing(QuickActionsTest.java:123)
```

**Comment by TabooSun** [comment]
I am able to reproduce this by setting up a personal Firebase project. The API 26 device (Samsung device) produces flakey test result for the `androidx.test.uiautomator.UiObjectNotFoundException` throughout my tests, the statistic is around 8 out of 10 that produces such error.

As for the `appShortcutLaunchActivityAfterPressing` test case, I am able to reproduce it on my own real device, Samsung Note 20 Ultra and I have patched it.

**Comment by stuartmorgan** [comment]
> It seems that `device.pressHome()` is not triggered in this particular device for some reason.

Your test appears to press home and then swipe immediately without waiting for any transition to home to complete. Is there a reason you believe the problem is with the test framework, rather than a race in your test code?

**Comment by TabooSun** [comment]
> Your test appears to press home and then swipe immediately without waiting for any transition to home to complete.

Tested with `device.waitForWindowUpdate(device.getLauncherPackageName(), 10000)`, still no luck.

> Is there a reason you believe the problem is with the test framework, rather than a race in your test code?

With 10000 of timeout, I don't think that it is due to race condition.

I believe the issue I am facing is the difference in the launchers. I am trying to figure out a way to reduce the discrepancies as much as possible. Any suggestion would be appreciated.

**Comment by stuartmorgan** [comment]
Looks like the tests are indeed passing in CI now! @camsim99 could you re-review the updated version?

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (camsim99) [inline packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java:1]: Accident?

**Review by camsim99 — APPROVED** [review]
LGTM! One nit, though

File: packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java
Comment (TabooSun) [reply packages/quick_actions/quick_actions_android/example/android/app/src/androidTest/java/io/flutter/plugins/quickactionsexample/QuickActionsTest.java]: Fixed.

**Comment by jld3103** [comment]
Awesome work! Really makes the UX a lot better.

**Pull request merged**
